From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 24 Jun 2023 18:56:46 +0200
Subject: [PATCH] Implement [TacoSpigot-0013] Disable Scoreboards for non
 players by default by Aikar

Entities collision is checking for scoreboards setting.
This is very heavy to do map lookups for every collision to check
this setting.

So avoid looking up scoreboards and short circuit to the "not on a team"
logic which is most likely to be true.

diff --git a/src/main/java/net/minecraft/server/CommandScoreboard.java b/src/main/java/net/minecraft/server/CommandScoreboard.java
index 27d08a7de54de062041a107490a9d4684872ec8b..18bd7c38b743791ec5358dca1cc2c83e177e5c99 100644
--- a/src/main/java/net/minecraft/server/CommandScoreboard.java
+++ b/src/main/java/net/minecraft/server/CommandScoreboard.java
@@ -471,6 +471,7 @@ public class CommandScoreboard extends CommandAbstract {
 
                     while (iterator.hasNext()) {
                         Entity entity = (Entity) iterator.next();
+                        if (!entity.world.tacoConfigUnicorn.allowNonPlayerEntitiesOnScoreboards && !(entity instanceof EntityHuman)) continue; // TacoSpigot
                         String s2 = e(icommandlistener, entity.getUniqueID().toString());
 
                         if (scoreboard.addPlayerToTeam(s2, s)) {
diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 44d3322367bcea7a60d60d437d63c4742731873d..c5518b03093737399dbd0d3728498f1091a19c89 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1821,6 +1821,7 @@ public abstract class EntityLiving extends Entity {
     }
 
     public ScoreboardTeamBase getScoreboardTeam() {
+        if (!this.world.tacoConfigUnicorn.allowNonPlayerEntitiesOnScoreboards && !(this instanceof EntityHuman)) return null; // TacoSpigot
         return this.world.getScoreboard().getPlayerTeam(this.getUniqueID().toString());
     }
 
diff --git a/src/main/java/org/mythicprojects/unicornspigot/config/TacoWorldConfig.java b/src/main/java/org/mythicprojects/unicornspigot/config/TacoWorldConfig.java
index b2cc18726a81d0aed7c65093750b2f11a653de61..61d4ff62b87efdeb8feb019e53574f01bd6e345d 100644
--- a/src/main/java/org/mythicprojects/unicornspigot/config/TacoWorldConfig.java
+++ b/src/main/java/org/mythicprojects/unicornspigot/config/TacoWorldConfig.java
@@ -18,6 +18,11 @@ public final class TacoWorldConfig extends EngineConfig.WorldConfig {
     @CustomKey("redstone-fire-BlockPhysicsEvent")
     public boolean redstoneFireBlockPhysicsEvent = true;
 
+    @Comment("Whether to allow non-players (animals, monsters etc.) to be on scoreboard.")
+    @Comment("Entities collision is checking for scoreboard setting, which is expensive especially if there are many entities.")
+    @Comment("Disabled by default since in most cases there is no need to have entities on scoreboard.")
+    public boolean allowNonPlayerEntitiesOnScoreboards = false;
+
     public Hopper hopper = new Hopper();
 
     public static class Hopper extends UnicornConfigSection {
